### PR TITLE
add lint

### PR DESCRIPTION
**Pull request by shiraji**
see #59 

# TODO

- [x] Add a lint that checks calling `@NeedsPermission` method, directly
- [x] Add a gradle task that injects lint.jar to PermissionsDispatcher

# NOT GOING TO DO

Because they make this pull request huge and there are no test cases, I won't do followings for this pull request.

* Add test cases
* Add a lint that checks implementing `onRequestPermissionsResult()` except WRITE_SETTINGS or SYSTEM_ALERT_WINDOW
* Add test cases for implementing `onRequestPermissionsResult()` except WRITE_SETTINGS or SYSTEM_ALERT_WINDOW
* Add a lint that checks implementing `onRequestPermissionsResult()` except WRITE_SETTINGS or SYSTEM_ALERT_WINDOW in Activity (not Fragment since user may skip it in case it handles in another Fragment.)
* Add test cases for implementing `onActivityResult()` if it's asking WRITE_SETTINGS or SYSTEM_ALERT_WINDOW in Activity

### CODE REVIEW

File: lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    @Override
+    protected boolean allowCompilationErrors() {
+        return true;
+    }
+
+    @Override
+    protected InputStream getTestResource(String relativePath, boolean expectExists) {
+        String path = "lint/src/test/java/permissions/dispatcher/data/".replace("/", File.separator) + relativePath;
```
Comment (shiraji) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: This line makes me crazy...I wish I find better way to get path to `data` directory. I follow what [android bundled lint does](https://android.googlesource.com/platform/tools/base/+/master/lint/libs/lint-tests/src/test/java/com/android/tools/lint/checks/AbstractCheckTest.java?autodive=0%2F%2F%2F%2F%2F%2F%2F#58)

Comment (aurae) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: I would suggest using `resources` for this, i.e. putting the referenced test data in the lint project's `src/test/resources` folder. Here's some example code on how to obtain the contents of the resource file `src/test/resources/test.txt`:

```java
String resource = "test.txt";
URL url = Thread.currentThread().getContextClassLoader().getResource(resource);
File file = new File(url.getPath());
return new FileInputStream(file);
```

Comment (shiraji) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: Nice! Let me fix that.

Comment (shiraji) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: Didn't work...not sure why. Always became `null`.

Comment (shiraji) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: And one more problem for this test case is the fact that `.class.data` is build by JDK7. It won't pass JDK8's test. Maybe this is the reason there is no test case in timber-lint.

**Comment by hotchemi** [comment]
Wow thx.

File: lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java (hunk 1 above)
Comment (shiraji) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: Someone uses [getTestResouce()](https://code.google.com/p/android/issues/detail?id=182195) to load resources. There are seems lots of problems in `LintDetectorTest`...

Comment (aurae) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: Which part of the above code results in `null`? Weird how `LintDetectorTest` doesn't want to play along...

Comment (shiraji) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: `Thread.currentThread().getContextClassLoader().getResource(resource);` ends up `null`. There is `resource` in build directory, but no luck at this moment.

Comment (aurae) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: You added the file to the build directory? It's supposed to go into `src/test/resources`, next to the `java` folder.

Comment (shiraji) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: No. I added files in `src/test/resources`. That automatically copy files in
`src/test/resources` which is expected. However, it is still `null`

2016年1月20日(水) 0:04 Marcel Schnelle <notifications@github.com>:

Comment (aurae) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: That's odd... I will check out the branch later and see for myself.

Comment (aurae) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: Okay, there really is something wrong about `LintDetectorTest`'s loading of test resources. On top of that, it is still stuck with JUnit 3 by extending `TestCase`... what do you think about using the [lint-junit-rule](https://github.com/a11n/lint-junit-rule) as a basis for our tests? It looks promising!

Comment (shiraji) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: Yea that's good idea. Let me convert my test cases. Thanks for researching!

Comment (aurae) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: On a side note, I'm running into the same issues again, even though the project claims to have fixed the issues with `getResourceAsStream()`.

Comment (shiraji) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: I don't think he/she uploaded to jcenter.

```groovy
    testCompile "de.ad:lint-junit-rule:0.3.0"
    testCompile 'de.ad:lint-assertj:0.3.0'
    testCompile 'de.ad:lint-hamcrest:0.3.0'
```

Could not resolve dependencies.

check here https://jcenter.bintray.com/de/ad/

Comment (aurae) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: Yeah, I had the same problem. I switched to pulling in the latest version of the dependency using Jitpack:

```gradle
allprojects {
    repositories {
        jcenter()
        maven { url "https://jitpack.io" }
    }
}
...
dependencies {
    testCompile "com.github.a11n.lint-junit-rule:lint-junit-rule:776b31de28"
    testCompile "com.github.a11n.lint-junit-rule:lint-assertj:776b31de28"
}
```

Comment (shiraji) [inline lint/src/test/java/permissions/dispatcher/PermissionDispatcherDetectorTest.java]: Nice. jitpack is always our friends! I will use that instead.

I posted [issue](https://github.com/a11n/lint-junit-rule/issues/14). See he can fix it.

**Comment by shiraji** [comment]
I posted another [issue](https://github.com/a11n/lint-junit-rule/issues/15) to lint-junit-rule. It didn't copy .java file to working directory.

**Comment by shiraji** [comment]
OK, after I talked with @a11n about `LintDetectorTest` and `lint-junit-rule`, I think there are 3 options for this issue.

1. Forget about tests (like Timber did)
1. Forget about Lint
1. Wait until someone improves `LintDetectorTest`

According to the [his comment](https://github.com/a11n/lint-junit-rule/issues/15#issuecomment-178527127), `lint-junit-rule` wraps `LintDetectorTest` but not extend it. So in order to run test correctly, we need to have .class/.java files which already did it at the beginning and failed to test it at travis ci.

Do you guys have any suggestion for this issue?

(To make it clear, `lint-junit-rule` is nothing wrong. I mean, we should use that library even though someone fixes `LintDetectorTest`)

**Comment by aurae** [comment]
Did you verify that your checks work, i.e. that they function correctly and show up in Lint reports? If so, I'd say let's do option 1.

**Comment by shiraji** [comment]
Yes. But I made a lot of mess. Let me clean up the branch first.

**Comment by shiraji** [comment]
## Report result

```xml
    <issue
        id="CallNeedsPermission"
        severity="Error"
        message="Trying to access permission-protected method directly."
        category="Correctness"
        priority="7"
        summary="A method annotated with @NeedsPermission should not be call directly"
        explanation="A method annotated with @NeedsPermission should not be call directly"
        errorLine1="        showCamera();"
        errorLine2="        ~~~~~~~~~~~~">
        <location
            file="/[DELETED]/PermissionDispatcherTest/app/src/main/java/com/github/shiraji/permissiondispatchertest/MainActivity.java"
            line="16"
            column="9"/>
    </issue>
```
[DELETED] is the path to `PermissionDispatcherTest` of "Hello world" project which I created as testing the .jar file.

MainActivity.java

```java
package com.github.shiraji.permissiondispatchertest;

import android.Manifest;
import android.support.v7.app.AppCompatActivity;

import permissions.dispatcher.NeedsPermission;
import permissions.dispatcher.RuntimePermissions;

@RuntimePermissions
public class MainActivity extends AppCompatActivity {
    @NeedsPermission(Manifest.permission.CAMERA)
    void showCamera() {
    }

    void foo() {
        showCamera();
    }
}
```

## Steps to run lint test

* clone [my repo](https://github.com/shiraji/PermissionsDispatcher)
* change branch to `59_add_lint`
* [OPTIONAL] change `gradle.perperties`'s VERSION to higher (e.g. 2.1.5)
* publish to local maven repo `./gradlew clean build publishToMavenLocal`
* create new project with above MainActivity.java
* add PermissionsDispatcher to build.gradle that use local maven's version
```gradle
allprojects {
    repositories {
        jcenter()
        mavenLocal()
    }
}

dependencies {
    compile fileTree(dir: 'libs', include: ['*.jar'])
    testCompile 'junit:junit:4.12'
    compile 'com.android.support:appcompat-v7:23.1.1'
    compile 'com.github.hotchemi:permissionsdispatcher:2.1.5'
    apk 'com.github.hotchemi:permissionsdispatcher-processor:2.1.5'
}
```
* run lint `./gradlew lint`

---

Do you have good **explanation** for this lint? I'm terrible for writing long explanation in English.

**Comment by shiraji** [comment]
Thanks. Squashed the revert commits.

**Comment by aurae** [comment]
The `message` portion already sounds great, but I'd remove the full stop at the end. Here's my proposal:

`message`
*Trying to access permission-protected method directly*

`summary`
*Call the corresponding "withCheck" method of the generated PermissionsDispatcher class instead*

`explanation`
*Directly invoking a method annotated with @NeedsPermission may lead to misleading behaviour on devices running Marshmallow and up. Therefore, it is advised to use the generated PermissionsDispatcher class instead, which provides a "withCheck" method that safely handles runtime permissions.*

**Comment by shiraji** [comment]
@aurae Super! Updated it. 

If this looks great, let me know. I will squash the commits.

**Comment by aurae** [comment]
LGTM

**Comment by hotchemi** [comment]
I’m gonna catch up!!!!!!!

**Comment by shiraji** [comment]
Squashed! Thanks for reviewing. 

Could you guys merge this?

**Comment by aurae** [comment]
Oh, just one more thing that I noticed only now: Your lint-related classes are missing the *s* in `PermissionsDispatcher`. Could you change those so that we're staying consistent? Other than that, great work!

**Comment by shiraji** [comment]
Sure. Nice catch.

**Comment by shiraji** [comment]
Fixed and squashed it.

**Comment by aurae** [comment]
Nice! Shintaro, I'll let you merge this after you have caught up and reviewed this. To me, it looks great, and it's a nice addition that puts PD apart from the plethora of other runtime permission libraries!

**Comment by shiraji** [comment]
The comment makes me feel proud of the work! 

Take your time, @hotchemi. If you have question, don't hesitate to ask me.

File: lint/src/main/java/permissions/dispatcher/PermissionsDispatcherDetector.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+public class PermissionsDispatcherDetector extends Detector implements Detector.JavaScanner {
+    public static final Issue ISSUE = Issue.create("CallNeedsPermission",
+            "Call the corresponding \"withCheck\" method of the generated PermissionsDispatcher class instead",
+            "Directly invoking a method annotated with @NeedsPermission may lead to misleading behaviour on devices running Marshmallow and up. Therefore, it is advised to use the generated PermissionsDispatcher class instead, which provides a \"withCheck\" method that safely handles runtime permissions.",
+            Category.CORRECTNESS,
+            7,
+            Severity.ERROR,
+            new Implementation(PermissionsDispatcherDetector.class, EnumSet.of(Scope.ALL_JAVA_FILES)));
+
+    static List<String> generatedClassNames = new ArrayList<>();
```
Comment (hotchemi) [inline lint/src/main/java/permissions/dispatcher/PermissionsDispatcherDetector.java]: Just one question, is it ok to write like below?

```java
static final List<String> generatedClassNames = new ArrayList<>();
```

**Comment by hotchemi** [comment]
@shiraji Sorry for the delay(I've been busy for developing droidkaigi app...).
I checked your commit and it's almost awesome! I just pointed out one thing and I'm gonna fix it after merged if it is ok. Anyway, really appreciate for your great work!

File: lint/src/main/java/permissions/dispatcher/PermissionsDispatcherDetector.java (hunk 2 above)
Comment (shiraji) [inline lint/src/main/java/permissions/dispatcher/PermissionsDispatcherDetector.java]: Yes. It can be final.

**Comment by shiraji** [comment]
NP. Thanks for merging this. I will work on another lint checks.

File: lint/src/main/java/permissions/dispatcher/PermissionsDispatcherDetector.java (hunk 2 above)
Comment (hotchemi) [inline lint/src/main/java/permissions/dispatcher/PermissionsDispatcherDetector.java]: OK.

**Comment by PawanDalal** [comment]
Why I am facing this `Error: Trying to access permission-protected method directly [CallNeedsPermission]?`

The lint complains this error and my build is failing due to this.
The error is reported in a Fragment which does not require any special permission and is not annotated with `@RuntimePermissions`.

This rule complains for a method in following scenario
1. Another method with same name in the Activity where this fragment is attached AND
2. The method in the activity is annotated with `@NeedsPermission({Manifest.permission.READ_SMS, Manifest.permission.RECEIVE_SMS})`

**Comment by shiraji** [comment]
@PawanDalal 

Could you raise an issue? This is the first PR for `CallNeedsPermission` lint however, the lint was rewritten at least 2 times after this PR. 

Also, could you give us sample code or minimal project that reproduce the problem in new issue?

**Pull request merged**